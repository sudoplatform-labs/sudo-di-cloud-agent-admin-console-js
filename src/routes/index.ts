export * from './route-definition';
export * from './utils';
