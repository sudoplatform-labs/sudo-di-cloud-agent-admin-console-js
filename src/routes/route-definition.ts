import { ImageName } from '../components/Image';
import { FeatureFlag } from '../feature-flags';

export interface RouteDefinition {
  /** Route path */
  path: string;

  /** Results in partial match if set to false (default = true) */
  exact?: boolean;

  /** Component to render for this route */
  component?: React.ComponentType;

  /** Child routes that are nested under this route */
  routes?: RouteDefinition[];

  /** Display name of route, used in Sidebar etc. */
  displayName?: string;

  /** Icon to be displayed in sidebar */
  iconName?: ImageName;

  /** Allow Icon classes to apply styling if necessary */
  iconClass?: string;

  /** Set to true if this route should appear in the sidebar */
  showInSidebar?: boolean;

  /** Restrict visibility based on a feature flag  */
  featureFlag?: FeatureFlag;
}

export type RouteDefinitionFactory = () => RouteDefinition;
