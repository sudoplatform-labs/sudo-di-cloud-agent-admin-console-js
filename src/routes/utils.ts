import { flatten, omit } from 'lodash';
import { matchPath } from 'react-router';
import { RouteDefinition } from './route-definition';

/** Flattens a heirarchy of routes into an array */
export function flattenRoutes(
  routes: RouteDefinition[],
  maxDepth?: number,
): RouteDefinition[] {
  const nextDepth = maxDepth && maxDepth - 1;

  return flatten(
    routes.map((route) => {
      const subRoutes =
        nextDepth === undefined || nextDepth > 1
          ? flattenRoutes(route.routes ?? [], nextDepth)
          : [];
      return [route, ...subRoutes];
    }),
  ).map((route) => omit(route, 'routes'));
}

/** Finds a route definition that matches a given path */
export function findRoute(
  urlPath: string,
  routes: RouteDefinition[],
): RouteDefinition | undefined {
  return routes.find(
    (route) =>
      route.exact !== false &&
      matchPath(urlPath, {
        path: route.path,
        exact: true,
      }),
  );
}

/** Returns all ancestor paths for `path`, including `path` itself */
export const getPathAncestry = (path?: string): string[] => {
  if (!path) return [];

  const pathSnippets = path.split('/').filter((snippet) => snippet);

  return pathSnippets.map((_, index) => {
    return `/${pathSnippets.slice(0, index + 1).join('/')}`;
  });
};
