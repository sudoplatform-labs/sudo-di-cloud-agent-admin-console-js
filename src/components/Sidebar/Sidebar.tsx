import { Menu } from 'antd';
import SubMenu from 'antd/lib/menu/SubMenu';
import color from 'color';
import React, { useCallback, useEffect, useState } from 'react';
import { Link, generatePath } from 'react-router-dom';
import styled from 'styled-components';
import { RouteDefinition, getPathAncestry } from '../../routes';
import { theme } from '../../theme';
import { Image } from '../Image';
import { SudoPlatformLogo } from '../SudoPlatformLogo';

const activeMenuBackgroundColor = color(theme.greys.midnight)
  .mix(color(theme.colors.sudoBlue), 0.3)
  .hex();

const activeMenuGroupBackgroundColor = color(theme.greys.midnight)
  .mix(color(theme.colors.sudoBlue), 0.03)
  .hex();

const StyledSudoPlatformLogo = styled(SudoPlatformLogo)`
  padding: 18px 0 12px;
`;

const StyledSidebar = styled.div`
  background-color: ${theme.greys.midnight};
  padding-bottom: 50px;
`;

const StyledMenu = styled(Menu)`
  &.ant-menu,
  .ant-menu {
    background-color: transparent;
    border-right: none;
    color: white;

    // Menu item containers
    .ant-menu-item,
    .ant-menu-submenu,
    .ant-menu-submenu-title {
      margin-top: 0;
      margin-bottom: 0;
      width: 100%;
      height: auto;
      padding-top: 0;
      padding-bottom: 0;
    }

    // Menuitem links
    a {
      display: block;
      margin-top: 8px;
      margin-bottom: 8px;
      line-height: 40px;
      color: white;
      text-decoration: none;
      cursor: pointer;
    }

    // Menu item hover state
    .ant-menu-submenu-active .ant-menu-submenu-title,
    .ant-menu-item-active {
      background-color: ${activeMenuBackgroundColor};
      color: white;
    }

    // Menu item selected state
    .ant-menu-item-selected {
      background-color: ${activeMenuBackgroundColor};

      &:after {
        border-right: 4px solid ${theme.colors.sudoBlue};
      }
    }

    .ant-menu-submenu-selected {
      background-color: ${activeMenuGroupBackgroundColor};
    }

    // Expand/contract icon
    .ant-menu-submenu,
    .ant-menu-submenu:hover {
      .ant-menu-submenu-arrow:before,
      .ant-menu-submenu-arrow:after {
        color: white;
        background: #fff;
      }
    }
  }

  // Force all icon images in the sidebar menu to a specific color
  // regardless on what the style guide differences might be.
  // NOTE : This assumes the icons are SVG format.
  img .colored-icon {
    fill: #ffffff;
  }
`;

const IconImage = styled(Image)`
  display: inline-block;
  margin-right: 12px;
`;

interface Props {
  pathName: string;
  routes: RouteDefinition[];
}

export const Sidebar = (props: Props): React.ReactElement => {
  const [openKeys, setOpenKeys] = useState(getPathAncestry(props.pathName));

  // Converts parameterized route paths to real paths, e.g.
  //   `/console/project/:projectId` => `/console/project/some-project`
  const getConcretePath = useCallback((routePath: string) => {
    return generatePath(routePath);
  }, []);

  // Function to filter out any routes we don't want to display
  const routeFilter = useCallback(
    (route: RouteDefinition) => route.showInSidebar,
    [],
  );

  // Close all open menus when current path changes
  useEffect(() => {
    setOpenKeys(getPathAncestry(props.pathName));
  }, [setOpenKeys, getConcretePath, props.pathName]);

  return (
    <StyledSidebar>
      <StyledSudoPlatformLogo name="sdca_light" width={100} />
      <StyledMenu
        mode="inline"
        onOpenChange={(keys) =>
          'keys' in keys ? setOpenKeys(keys as string[]) : null
        }
        selectedKeys={[getConcretePath(props.pathName)]}
        openKeys={openKeys}>
        {props.routes.filter(routeFilter).map((route) => {
          const concretePath = getConcretePath(route.path);
          return route.routes ? (
            <SubMenu
              key={concretePath}
              title={
                <Link to={concretePath}>
                  {route.iconName && (
                    <IconImage
                      name={route.iconName}
                      className={route.iconClass}
                    />
                  )}
                  {route.displayName}
                </Link>
              }>
              {route.routes.filter(routeFilter).map((subRoute) => {
                const concreteChildPath = getConcretePath(subRoute.path);
                return (
                  <Menu.Item key={concreteChildPath}>
                    <Link to={concreteChildPath}>
                      {subRoute.iconName && (
                        <IconImage
                          name={subRoute.iconName}
                          className={route.iconClass}
                        />
                      )}
                      {subRoute.displayName}
                    </Link>
                  </Menu.Item>
                );
              })}
            </SubMenu>
          ) : (
            <Menu.Item key={concretePath}>
              <Link to={concretePath}>
                {route.iconName && (
                  <IconImage
                    name={route.iconName}
                    className={route.iconClass}
                  />
                )}
                {route.displayName}
              </Link>
            </Menu.Item>
          );
        })}
      </StyledMenu>
    </StyledSidebar>
  );
};
