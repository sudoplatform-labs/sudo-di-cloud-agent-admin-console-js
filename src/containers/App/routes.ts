import { createRouteRedirect } from '../../components/RouteRenderer';
import { RouteDefinitionFactory } from '../../routes';
import { Console } from '../Console';

export const getRoot: RouteDefinitionFactory = () => ({
  path: '/',
  component: createRouteRedirect('/console'),
  routes: [
    {
      path: '/console',
      component: Console,
      exact: false,
    },
  ],
});
