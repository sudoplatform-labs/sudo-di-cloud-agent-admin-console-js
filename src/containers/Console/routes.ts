import { RouteDefinitionFactory } from '../../routes';
import * as Pages from '../../pages';
import { createRouteRedirect } from '../../components/RouteRenderer';

export const getConsoleRoutesRoot: RouteDefinitionFactory = () => ({
  path: '/console',
  component: createRouteRedirect('/console/dashboard'),
  routes: [
    {
      path: '/console/dashboard',
      component: Pages.ConsoleDashboard,
      displayName: 'Cloud Agent Dashboard',
      iconName: 'home',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/settings',
      component: Pages.Settings,
      displayName: 'Settings',
      iconName: 'settings',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/decentralized-identifiers',
      component: Pages.DecentralizedIdentifiersDashboard,
      displayName: 'Decentralized Identifiers',
      iconName: 'icon-did-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/connections',
      component: Pages.ConnectionsDashboard,
      displayName: 'Connections',
      iconName: 'icon-connection-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
    },
    {
      path: '/console/holder-wallet',
      component: createRouteRedirect('/console/holder-wallet/credentials'),
      displayName: 'Holder Wallet',
      iconName: 'icon-wallet-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
      routes: [
        {
          path: '/console/holder-wallet/credentials',
          component: Pages.Credentials,
          displayName: 'My Credentials',
          showInSidebar: true,
        },
        {
          path: '/console/holder-wallet/proofs',
          component: Pages.PresentedProofs,
          displayName: 'My Proof Presentations',
          showInSidebar: true,
        },
      ],
    },
    {
      path: '/console/credential-issuer',
      component: createRouteRedirect('/console/credential-issuer/dashboard'),
      displayName: 'Credential Issuer',
      iconName: 'icon-issuer-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
      routes: [
        {
          path: '/console/credential-issuer/dashboard',
          component: Pages.CredentialIssuerDashboard,
          displayName: 'Dashboard',
          showInSidebar: true,
        },
        {
          path: '/console/credential-issuer/credential-catalogue',
          component: Pages.CredentialCatalogue,
          displayName: 'Catalogue',
          showInSidebar: true,
        },
        {
          path: '/console/credential-issuer/credential-requests',
          component: Pages.CredentialIssuance,
          displayName: 'Credential Issuance',
          showInSidebar: true,
        },
      ],
    },
    {
      path: '/console/credential-verifier',
      component: createRouteRedirect('/console/credential-verifier/dashboard'),
      displayName: 'Credential Verifier',
      iconName: 'icon-verifier-line',
      iconClass: 'colored-icon',
      showInSidebar: true,
      routes: [
        {
          path: '/console/credential-verifier/dashboard',
          component: Pages.CredentialVerifierDashboard,
          displayName: 'Dashboard',
          showInSidebar: true,
        },
        {
          path: '/console/credential-verifier/proofs',
          component: Pages.RequestedProofs,
          displayName: 'Proof Requests',
          showInSidebar: true,
        },
      ],
    },
  ],
});
