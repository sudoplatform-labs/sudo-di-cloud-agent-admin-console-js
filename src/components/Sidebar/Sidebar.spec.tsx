import { Menu } from 'antd';
import { shallow } from 'enzyme';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { Link } from 'react-router-dom';
import { RouteDefinition } from '../../routes';
import { acwMount } from '../../test/enzyme';
import { Sidebar } from './Sidebar';

const TestComponent: React.FC<{}> = () => <div />;

const routes: RouteDefinition[] = [
  {
    component: TestComponent,
    path: '/aaa',
    showInSidebar: true,
    displayName: 'AAA',
  },
  {
    component: TestComponent,
    path: '/bbb',
    showInSidebar: true,
    displayName: 'BBB',
  },
  {
    component: TestComponent,
    path: '/ccc',
    showInSidebar: true,
    displayName: 'CCC',
  },
  {
    path: '/ddd',
    showInSidebar: true,
    displayName: 'CCC',
    iconName: 'home',
    routes: [
      {
        component: TestComponent,
        path: '/ddd/ddd-1',
        showInSidebar: true,
        displayName: 'DDD-1',
      },
      {
        component: TestComponent,
        path: '/ddd/ddd-2',
        showInSidebar: true,
        displayName: 'DDD-1',
      },
    ],
  },
];

describe('Sidebar', () => {
  it('should render', () => {
    const wrapper = shallow(<Sidebar pathName="/dashboard" routes={routes} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all items', async () => {
    const wrapper = await acwMount(
      <Sidebar pathName="/dashboard" routes={routes} />,
    );
    wrapper.update();

    const hasRoute = (path: string): boolean =>
      wrapper.findWhere((node) => node.prop('to') === path).length > 0;

    expect(hasRoute('/aaa')).toBeTruthy();
    expect(hasRoute('/bbb')).toBeTruthy();
    expect(hasRoute('/ccc')).toBeTruthy();
  });

  it('should load with properly expanded submenus', async () => {
    const wrapper = await acwMount(
      <Sidebar pathName="/ddd/ddd-1" routes={routes} />,
    );

    expect(wrapper.find(Menu).prop('openKeys')).toEqual(['/ddd', '/ddd/ddd-1']);
  });

  it('should change menuitem', async () => {
    const wrapper = await acwMount(<Sidebar pathName="/aaa" routes={routes} />);

    expect(wrapper.find(Menu).prop('openKeys')).toEqual(['/aaa']);

    await act(async () => {
      const ddd1Link = wrapper
        .find(Link)
        .filterWhere((l) => l.prop('to') === '/ddd');
      ddd1Link.simulate('click');
    });
    wrapper.update();

    expect(wrapper.find(Menu).prop('openKeys')).toEqual(['/aaa', '/ddd']);
  });

  it('should not close submenu when user clicks on a child menuitem', async () => {
    const wrapper = await acwMount(
      <Sidebar pathName="/ddd/ddd-2" routes={routes} />,
    );

    expect(wrapper.find(Menu).prop('openKeys')).toEqual(['/ddd', '/ddd/ddd-2']);

    await act(async () => {
      const ddd2Link = wrapper
        .find(Link)
        .filterWhere((l) => l.prop('to') === '/ddd/ddd-2');
      ddd2Link.simulate('click');
    });
    wrapper.update();

    expect(wrapper.find(Menu).prop('openKeys')).toEqual(['/ddd', '/ddd/ddd-2']);
  });

  it('should not expand submenus', async () => {
    const wrapper = await acwMount(
      <Sidebar pathName="/ddd/ddd-2" routes={routes} />,
    );

    expect(wrapper.find(Menu).prop('openKeys')).toEqual(['/ddd', '/ddd/ddd-2']);
  });
});
