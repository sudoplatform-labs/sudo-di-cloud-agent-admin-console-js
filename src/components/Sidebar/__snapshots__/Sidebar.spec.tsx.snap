// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should render 1`] = `
<styled.div>
  <Styled(SudoPlatformLogo)
    name="sdca_light"
    width={100}
  />
  <Styled(Menu)
    mode="inline"
    onOpenChange={[Function]}
    openKeys={
      Array [
        "/dashboard",
      ]
    }
    selectedKeys={
      Array [
        "/dashboard",
      ]
    }
  >
    <MenuItem
      key="/aaa"
    >
      <Link
        to="/aaa"
      >
        AAA
      </Link>
    </MenuItem>
    <MenuItem
      key="/bbb"
    >
      <Link
        to="/bbb"
      >
        BBB
      </Link>
    </MenuItem>
    <MenuItem
      key="/ccc"
    >
      <Link
        to="/ccc"
      >
        CCC
      </Link>
    </MenuItem>
    <SubMenu
      key="/ddd"
      title={
        <ForwardRef
          to="/ddd"
        >
          <ForwardRef(Styled(Image))
            name="home"
          />
          CCC
        </ForwardRef>
      }
    >
      <MenuItem
        key="/ddd/ddd-1"
      >
        <Link
          to="/ddd/ddd-1"
        >
          DDD-1
        </Link>
      </MenuItem>
      <MenuItem
        key="/ddd/ddd-2"
      >
        <Link
          to="/ddd/ddd-2"
        >
          DDD-1
        </Link>
      </MenuItem>
    </SubMenu>
  </Styled(Menu)>
</styled.div>
`;
